### PR TITLE
Match upstream API changes for IndexFormat

### PR DESCRIPTION
**Pull request by kejor**
**Connections**
#978 

**Description**
Match upstream API changes for IndexFormat.

- `index_format` from `VertexStateDescriptor` is now a `Option`
- `RenderCommand::SetIndexBuffer` now has a `index_format` attribute
- The index format for execution is now set when matched both `RenderCommand::SetPipeline` and `RenderCommand::SetIndexBuffer`.

**Testing**
Tested with the examples of https://github.com/kejor/wgpu-rs/commit/e5f82a21972842806774629534ab97051043489d

---

Should I implement this?https://github.com/gpuweb/gpuweb/blob/0855ad9b97d4fa6a8cb4d51740675a21044f099c/spec/index.bs#L3703

### CODE REVIEW

File: player/tests/data/quad.ron
Comment (kvark) [inline player/tests/data/quad.ron:89]: this should actually be `None`. We only need to specify this for strip topologies

File: wgpu-core/src/command/render.rs
Comment (kvark) [inline wgpu-core/src/command/render.rs:1179]: We actually need to track both the `Option<IndexFormat>` from the current pipeline and the current index format of the active `SetIndexBuffer`. Before issuing a draw call there is an `is_ready` function that needs to check for:
  - if the pipeline index format is set, it has to be the same as the current index format, otherwise an error is generated

File: wgpu-core/src/command/bundle.rs
Comment (kvark) [inline wgpu-core/src/command/bundle.rs:213]: we also need to call `state.index.update_limit();` again if this format is changed here

File: wgpu-types/src/lib.rs
Comment (kvark) [inline wgpu-types/src/lib.rs:947]: Curious why this change was needed?

**Review by kvark — COMMENTED** [review]
Thank you for taking a stab at this! There is a few things we need to fix before landing. Also, please squash the commits as our merger bot doesn't know how to do it.

**Comment by kvark** [comment]
We are in a situation where the group may back-track from the decision to have the index format here at all. Let's hold this off until the group figures out what to do. Sorry about putting you in the middle of this!

**Comment by kvark** [comment]
I did some experiments in https://github.com/gpuweb/gpuweb/issues/1220#issuecomment-732357591, discussed it with the group, and it sounds like we can proceed with the API as it today. This PR is therefore unlocked. Please rebase!

**Comment by kvark** [comment]
@kejor would you be interested in finishing this? Sorry about blocking progress here previously!

**Comment by kejor** [comment]
> @kejor would you be interested in finishing this? Sorry about blocking progress here previously!

Yeah, no problemo. I was about to get started haha

File: wgpu-types/src/lib.rs
Comment (kejor) [reply wgpu-types/src/lib.rs:966]: There was one or two checks complaining about it. I don't quite remember which ones.

File: wgpu-core/src/command/bundle.rs
Comment (kejor) [reply wgpu-core/src/command/bundle.rs:534]: `execute` in `RenderBundle` doesn't really have a state, should it?

**Review by kvark — APPROVED** [review]
I think it's all good now, thank you for finishing this!
bors r+

**Comment by kvark** [comment]
@kejor would you be able to make a `wgpu-rs` update that adjusts for these API changes?

**Comment by kejor** [comment]
> @kejor would you be able to make a `wgpu-rs` update that adjusts for these API changes?

Yes, I'll be making a PR with the update.

**Pull request merged**
